Include Image Controls in accessibility tree

yoelhawa opened the pull request:
#### 6c9721632fc5a1379b9e26ad416d22e37c67f7d4
<pre>
Include Image Controls in accessibility tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=304218">https://bugs.webkit.org/show_bug.cgi?id=304218</a>
<a href="https://rdar.apple.com/166549601">rdar://166549601</a>

Reviewed by Joshua Hoffman.

This change makes images with controls return NO from
isAccessibilityElement, causing the controls to appear as children in
the accessibility tree. To preserve the image&apos;s alt text accessibility,
we create a mock UIAccessibilityElement that copies the image&apos;s
properties, then prepend it to the children array at index 0.

VoiceOver can now navigate to both the image content (via the mock
element) and the interactive controls.

* Source/WebCore/Configurations/WebCore.xcconfig:
Added conditional UIKit framework linking when
ENABLE_SPATIAL_IMAGE_CONTROLS is defined.

* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
Added m_cachedMockImageElement to cache the mock element.

* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
Added UIAccessibilityElement forward declaration and interface.
(mockImageElement):
Creates and caches mock element with image properties.
(determineIsAccessibilityElement):
Return NO for images with controls.
(accessibilityElements):
Prepend mock element to children array.
(accessibilityElementCount):
(accessibilityElementAtIndex):
(indexOfAccessibilityElement):
Mock element is at index 0, other indices shifted by +1.
(detach):
(_clearCachedIsAccessibilityElementState):
Clear cached mock element.

Canonical link: <a href="https://commits.webkit.org/305315@main">https://commits.webkit.org/305315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6bf783367a7e6844a936df507436058e7e7a003c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146141 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/91044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139942 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11136 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10586 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/105600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/91044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/141015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8318 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/123792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86444 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/7931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5691 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6422 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/117329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41988 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148852 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10116 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42544 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/113998 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10133 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114331 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29055 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7862 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120075 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/64840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10162 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/38027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9893 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/73730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10103 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9954 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -933,6 +937,17 @@ - (BOOL)determineIsAccessibilityElement
+        auto* node = backingObject->node();
+        if (auto* imageElement = dynamicDowncast<HTMLImageElement>(node)) {
+            if (SpatialImageControls::hasSpatialImageControls(*imageElement))
+                return false;
```
If we return `false` here, Images will no longer be exposed to ATs (like VoiceOver) on iOS. So, if an image has an `alt` tag, a user won't be able to land on the image to hear that description.

We may want to consider alternatives for exposing the children here, such as returning them as siblings of the image.

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -233,6 +239,36 @@ - (TextMarkerData)textMarkerData
+
+    if (RefPtr imageElement = dynamicDowncast<HTMLImageElement>(backingObject->node())) {
+        if (SpatialImageControls::hasSpatialImageControls(*imageElement)) {
+            auto* cache = backingObject->axObjectCache();
```
I think this should be a CheckedPtr since we call a non-trivial function with it (`getOrCreate`)

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -233,6 +239,36 @@ - (TextMarkerData)textMarkerData
+            NSMutableArray *views = [NSMutableArray array];
+
+            for (auto& child : composedTreeChildren(*imageElement)) {
+                if (RefPtr childObject = cache->getOrCreate(&child))
```
`getOrCreate` can take a reference rather than a pointer, avoiding a null-check. Can we remove the `&` from `&child`?

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -233,6 +239,36 @@ - (TextMarkerData)textMarkerData
+    if (![self _prepareAccessibilityCall])
+        return nil;
+
+    auto* backingObject = self.axBackingObject;
```
I think this should be a RefPtr to ensure the object stays alive for the duration of the function.

hoffmanjoshua commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -259,7 +277,38 @@ - (void)detach
+}
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+- (id)imageControlsMockElement
```
This element represents the image (rather than the controls themselves, which are a different child), right? If that's the case, do you think we could name this something like `mockImageElement`, or another name if you have one? On first read, this might get misunderstood to be the controls themselves.

hoffmanjoshua reviewed (approved):
This looks good to me! Thanks for making this change

Commit:
Include Image Controls in accessibility tree
https://bugs.webkit.org/show_bug.cgi?id=304218
rdar://166549601

Reviewed by Joshua Hoffman.

This change makes images with controls return NO from
isAccessibilityElement, causing the controls to appear as children in
the accessibility tree. To preserve the image's alt text accessibility,
we create a mock UIAccessibilityElement that copies the image's
properties, then prepend it to the children array at index 0.

VoiceOver can now navigate to both the image content (via the mock
element) and the interactive controls.

* Source/WebCore/Configurations/WebCore.xcconfig:
Added conditional UIKit framework linking when
ENABLE_SPATIAL_IMAGE_CONTROLS is defined.

* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
Added m_cachedMockImageElement to cache the mock element.

* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
Added UIAccessibilityElement forward declaration and interface.
(mockImageElement):
Creates and caches mock element with image properties.
(determineIsAccessibilityElement):
Return NO for images with controls.
(accessibilityElements):
Prepend mock element to children array.
(accessibilityElementCount):
(accessibilityElementAtIndex):
(indexOfAccessibilityElement):
Mock element is at index 0, other indices shifted by +1.
(detach):
(_clearCachedIsAccessibilityElementState):
Clear cached mock element.

Canonical link: https://commits.webkit.org/305315@main
